### PR TITLE
Add flake detection mode to `patrol drive`

### LINKED ISSUE #519: Introduce "flake detection" mode
Alternative title: **make it faster to run the same test multiple times**

### The problem

Flaky tests are bad – we know it. The standard way to detect them is to run the test many times. Currently it can be done in Patrol by simply running `patrol_cli` in the loop – we show our users how to do it in the [Tips & Tricks section in the docs](https://patrol.leancode.co/tips-and-tricks#running-a-test-many-times).

This works, but the performance could be much better. We rebuild the app on every test run, even though no changes are made.

### Suggested solution

We could build the app once, before the first test run, and then reuse that build in the subsequent test runs.

**Goals:**
- add the optional `-n <int>` argument to `patrol drive`, which specifies the number of times a test run should be repeated
  - the `-n` flag can be used only when a single `--target` is given. Otherwise, we print a friendly error message ("flake detection mode is not yet supported for multiple targets") and exit with code 1.

We could possibly allow for multiple `--targets`, but this will make the implementation more complex. We'd also have to decide on the order in which to build and run the tests. I think this can be postponed for later, especially since I feel that usually, our users will try to detect flakiness in a single test at once, not multiple. Also, I don't see any big benefits it'd have over calling (the hypothetical) `patrol drive -t <target> -n <int>` in a loop (just like we do currently).

**Non-goals:**
- detect if any files were changed during the test runs

  We don't care about this, it's the user's responsibility.
- report the results in a nice way

  This is important but I feel it fits in a separate issue/PR pair.

cc: @jBorkowska @jakubfijalkowski @mateuszwojtczak @shilangyu – I'd love to know what you think about this feature.

### PR DESCRIPTION
**Pull request by bartekpacia**
Fixes #519 

Checked
- example app (✅Android,✅iOS)
- customer: budowa (Android)

### CODE REVIEW

File: packages/patrol_cli/lib/src/features/drive/flutter_driver.dart [hunk 1]
```diff
@@ -161,20 +243,31 @@ class FlutterDriver {
       }
     }
 
+    String getApplicationBinaryPath() {
+      if (platform == TargetPlatform.android) {
+        const prefix = 'build/app/outputs/flutter-apk';
+        if (flavor != null) {
+          return join(prefix, 'app-${flavor.toLowerCase()}-debug.apk');
```
Comment (bartekpacia) [inline packages/patrol_cli/lib/src/features/drive/flutter_driver.dart:250]: hacky hacky

but I don't know how to do it better

File: packages/patrol_cli/lib/src/features/drive/drive_command.dart [hunk 2]
```diff
@@ -166,6 +175,22 @@ class DriveCommand extends StagedCommand<DriveCommandConfig> {
...
 
+    var repeat = 1;
+    try {
+      final repeatStr = argResults?['repeat'] as String? ?? '1';
+      repeat = int.parse(repeatStr);
+    } on FormatException {
+      throw const FormatException('`repeat` argument is not an int');
+    }
+
+    if (repeat != 1) {
```
Comment (shilangyu) [inline packages/patrol_cli/lib/src/features/drive/drive_command.dart:186]: Missing check for negative repeats

File: packages/patrol_cli/lib/src/features/drive/flutter_tool.dart [hunk 3]
```diff
@@ -50,22 +72,84 @@ class FlutterDriver {
...
+    );
+
+    int? exitCode;
+    final process = await Process.start(
+      'flutter',
+      [
+        '--no-version-check',
+        ...['build', platform.command],
+        '--debug',
+        ...['--target', target],
```
Comment (shilangyu) [inline packages/patrol_cli/lib/src/features/drive/flutter_tool.dart:98]: I like this way of organizing

File: packages/patrol_cli/lib/src/features/drive/flutter_tool.dart [hunk 4]
```diff
@@ -50,22 +72,84 @@ class FlutterDriver {
...
+        if (flavor != null) ...['--flavor', flavor!],
+        if (platform == TargetPlatform.iOS)
+          if (device.real) '--no-simulator' else '--simulator',
+        for (final dartDefine in {...env, ...dartDefines}.entries) ...[
+          '--dart-define',
+          '${dartDefine.key}=${dartDefine.value}',
+        ]
+      ],
+    );
+    String kill() {
```
Comment (shilangyu) [inline packages/patrol_cli/lib/src/features/drive/flutter_tool.dart:108]: newline

**Review by shilangyu — APPROVED** [review]

File: packages/patrol_cli/lib/src/features/drive/drive_command.dart (hunk 2 above)
Comment (bartekpacia) [reply packages/patrol_cli/lib/src/features/drive/drive_command.dart:190]: Right, fixed.

**Pull request merged**